### PR TITLE
Fix the retry math

### PR DESCRIPTION
**Pull request by szmarczak**
Fixes a bug where the retry delay is calculated incorrectly.

@timdp @UltCombo

### CODE REVIEW

File: source/request-as-event-emitter.js [hunk 1]
```diff
@@ -227,7 +226,7 @@ module.exports = (options, input) => {
 		let backoff;
 
 		try {
-			backoff = options.retry.retries(++retryTries, error);
+			backoff = options.retry.retries(retryCount, error);
```
Comment (FabricioMatteMassive) [inline source/request-as-event-emitter.js]: This is a breaking change:

The client can provide a custom `retries` function and it expects to receive `1` for the `iteration` argument in the first retry. This breaks the contract, it's a SEMVER-MAJOR change. I'd prefer to keep the old behavior.

**Comment by UltCombo** [comment]
Oh I've submitted the review with my work account, oops. :sweat_smile:

File: source/normalize-arguments.js [hunk 2]
```diff
@@ -236,7 +236,7 @@ const normalize = (url, options, defaults) => {
 			}
 
 			const noise = Math.random() * 100;
-			return ((1 << iteration) * 1000) + noise;
+			return ((2 ** iteration) * 1000) + noise;
```
[the remark this answers is not in the document]
Comment (sindresorhus) [reply source/normalize-arguments.js]: Yes, https://github.com/sindresorhus/got/blob/50fdab303cb2a6b34383de13e5e0ece9ceaf80c9/package.json#L9

**Comment by sindresorhus** [comment]
@szmarczak Can you include a description in the PR body and add a test?

File: source/request-as-event-emitter.js (hunk 1 above)
Comment (sindresorhus) [reply source/request-as-event-emitter.js]: > The client can provide a custom retries function and it expects to receive 1 for the iteration argument in the first retry.

@szmarczak We need a test for this.

Comment (sindresorhus) [reply source/request-as-event-emitter.js]: Yes, we need to preserve backwards compatibility. We can add it to https://github.com/sindresorhus/got/issues/639 if there's any benefit to it.

Comment (szmarczak) [reply source/request-as-event-emitter.js]: > The client can provide a custom retries function and it expects to receive 1 for the iteration argument in the first retry. This breaks the contract, it's a SEMVER-MAJOR change. I'd prefer to keep the old behavior.

Actually, no. It's a bug. According to the docs, it should start with 0:

> Delays between retries counts with function 1000 * Math.pow(2, retry) + Math.random() * 100, where retry is attempt number (starts from 0).

Comment (szmarczak) [reply source/request-as-event-emitter.js]: @sindresorhus 

> we need to preserve backwards compatibility

Compatibility with the docs or compatibility with the previous mechanism?

Comment (UltCombo) [reply source/request-as-event-emitter.js]: @szmarczak I see it as explaining the default logic, not the actual arguments. I'm using a custom `retries` function so it's a breaking change.

The documentation should also be updated to document what arguments the `retries` function gets.

Comment (szmarczak) [reply source/request-as-event-emitter.js]: In case we agree to save the previous mechanism, then yes.

@sindresorhus any thoughts?

Comment (sindresorhus) [reply source/request-as-event-emitter.js]: > @sindresorhus any thoughts?

Yes, like I said, backwards-compatibility must be preserved.

Comment (szmarczak) [reply source/request-as-event-emitter.js]: Right. Then it starts from 1.

Comment (sindresorhus) [reply source/request-as-event-emitter.js]: @szmarczak But if you think it should be changed, we can do so in Got 10. Add a comment https://github.com/sindresorhus/got/issues/639 if so.

Comment (UltCombo) [reply source/request-as-event-emitter.js]: I generally prefer 0-indexed collections, however in this specific case I believe it makes sense that `iteration === 1` in the first retry attempt. Having retry attempt as 0 is not intuitive.

Having the first retry attempt as 1 allows writing semantic code like:

```js
got('url', {
  retry: {
    // Retry up to 2 times.
    retries: retry => retry <= 2 ? getRetryDelay() : 0,
  },
});
```

@sindresorhus @szmarczak

**Pull request merged**